### PR TITLE
Merging reading from archives #50321

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Merging https://github.com/ClickHouse/ClickHouse/pull/50321

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/IO/Archives/LibArchiveReader.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+{
+    if (!dynamic_cast<FileEnumeratorImpl *>(enumerator.get()))
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Wrong enumerator passed to readFile()");
+    auto enumerator_impl = std::unique_ptr<FileEnumeratorImpl>(static_cast<FileEnumeratorImpl *>(enumerator.release()));
+    auto handle = std::move(*enumerator_impl).releaseHandle();
+    return std::make_unique<ReadBufferFromLibArchive>(std::move(handle), path_to_archive);
+}
+
+template <typename ArchiveInfo> std::unique_ptr<typename LibArchiveReader<ArchiveInfo>::FileEnumerator>
+LibArchiveReader<ArchiveInfo>::nextFile(std::unique_ptr<ReadBuffer> read_buffer)
```
[the remark this answers is not in the document]
Comment (antonio2368) [reply src/IO/Archives/LibArchiveReader.cpp:297]: yes I can do it that way, I agree it's a bit ugly

File: src/IO/Archives/LibArchiveReader.cpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    {
+        other.current_archive = nullptr;
+        other.current_entry = nullptr;
+    }
+
+    ~Handle()
+    {
+        if (current_archive)
+        {
+            archive_read_close(current_archive);
```
[the remark this answers is not in the document]
Comment (antonio2368) [reply src/IO/Archives/LibArchiveReader.cpp:45]: Actually now I see `archive_entry_clear` and `archive_entry_free` which I would need to use. I had the same comment in the original PR but forgot to take a closer look again. Good catch!

[the remark this answers is not in the document]
Comment (antonio2368) [reply src/IO/Archives/LibArchiveReader.cpp:45]: And now I found the part why I decided that there is no need for manually handling entry:
```
     archive_read_next_header()
             Read the header for the next entry and return a pointer to a struct archive_entry.
             This is a convenience wrapper around archive_read_next_header2() that reuses an
             internal struct archive_entry object for each request.
 ```
 
 so entry is handled internally, there is `archive_read_next_header2` for manually handling of entry

File: src/Storages/StorageFile.cpp [hunk 3]
```diff
@@ -1553,4 +1722,32 @@ void StorageFile::addColumnsToCache(
     schema_cache.addMany(cache_keys, columns);
 }
 
+void StorageFile::parseFileSource(String source, String & filename, String & path_to_archive)
+{
+    size_t pos = source.find("::");
```
[the remark this answers is not in the document]
Comment (antonio2368) [reply src/Storages/StorageFile.cpp:1727]: didn't see many other options that support reading from archive so not sure what is a common syntax
simply picked up the syntax that the original author decided on

**Review by vdimir — APPROVED** [review]

**Pull request merged**